pulp_python/app/pypi/views.py

```diff
@@ -78,11 +91,83 @@ def get_drvc(self, path):
         return distro, repo_ver, content
 
 
-class SimpleView(ViewSet, PyPIMixin):
+class PackageUploadMixin(PyPIMixin):
+    """A Mixin to provide package upload support."""
+
+    def upload(self, request, path):
+        """
+        Upload a package to the index.
+
+        0. Check request is in correct format
+        1. Check if the index allows uploaded packages (live-api-enabled)
+        2. Check if the package is in the repository already
+        3. If present then reject request
+        4. Spawn task to add content if no/old session present
+        5. Add uploads to current session to group into one task
+        """
+        distro = self.get_distribution(path)
+        if not distro.allow_uploads:
+            return HttpResponseForbidden(reason="Index is not editable")
+
+        repo = distro.repository
+        if not repo:
+            return HttpResponseBadRequest(reason="Index is not pointing to a repository")
+
+        serializer = PackageUploadSerializer(data=request.data)
+        serializer.is_valid(raise_exception=True)
+        artifact, filename = serializer.validated_data["content"]
+        repo_content = self.get_content(self.get_repository_version(distro))
+        if repo_content.filter(filename=filename).exists():
+            return HttpResponseBadRequest(reason=f"Package {filename} already exists in index")
+
+        return self.upload_package(repo, artifact, filename, request.session)
+
+    def upload_package(self, repo, artifact, filename, session):
+        """Steps 4 & 5, spawns tasks to add packages to index."""
+        start_time = datetime.now(tz=timezone.utc) + timedelta(seconds=5)
+        task = "updated"
+        if not session.get("start"):
+            task = self.create_upload_task(session, repo, artifact, filename, start_time)
+        else:
+            sq = Session.objects.select_for_update(nowait=False).filter(pk=session.session_key)
```

Comment:
This should not wait if someone else has the lock on this object. Possible `nowait` param should be `True`.